### PR TITLE
Album size

### PR DESCRIPTION
**Pull request by johnyenter-briars**
Greetings! 

#### Thank you!
First off, thank you for all the work you have put into spotify-qt! It is a great piece of software, and I use it on the daily without any issue.

Up front: I'm opening this PR is more of a "spit-balling" PR, as I attempted to implement the feature on my own, but I feel like this may not be optimal, nor necessarily the direction you would want to take this feature.

#### Objective

I feel like spotify-qt would benefit from a larger (and possible scalable) album art widget on the bottom left corner. Album art is so fundamentally connected with music consumption, I feel that displaying the album art in a larger/clear UI would enhance the experience.

#### First Attempt

The code updated in this PR adds a bunch of `setAlbum` overloads to support getting the image data back as a straight string, then parsing that to a pixmap.

Which I think is the right size, but certainly not good. Especially given that at this scale, the pixmap is already badly pixelated. (Attached image as example)
update_album_size

#### A Better Way?

I noticed that the logic in `cover.cpp` does some scaling of the image that I couldn't quite replicate. Does that logic produce the sharper image of the artist?

Do you have any general throught on this and/or does this direction make sense/needed?

Would appreciate any and all feedback. Thanks for reading!

### CODE REVIEW

**Comment by kraxarn** [comment]
Thank you :relaxed:

While I'm not against the idea, I definitively think it should be an option to use either large or small album art. It should probably also still show the name of the track and artist, maybe underneath? The image itself is only small as that's the only sized fetched. It's also available in 300×300 and 640×640 sizes, it's just never parsed or cached:
https://github.com/kraxarn/spotify-qt/blob/f3d6d34f36bf81bccfc76208daad782c1d953fca/lib/src/spotify/track.cpp#L113
It probably needs to be fixed before this merges though xd, but that's something I can look into.

The scaling for the artist is mostly to keep the same height of the image when adjusting the width of the side panel. Though, that also crops the image, which is fine for artist covers, but doesn't make sense for album covers.

File: .gitignore
Comment (kraxarn) [inline .gitignore:219]:
```suggestion

# VS Code
.vscode/
```
This makes it look like it's a part of "Idea based stuff".

File: src/mainwindow.cpp
Comment (kraxarn) [inline src/mainwindow.cpp:423]: Any specific reason for switching this out to a string?

File: src/view/context/content.hpp
Comment (kraxarn) [inline src/view/context/content.hpp:38]: The small size should probably still be a fixed 64×64. The large one should be dynamic to whatever the width of the side panel is, and wouldn't need a fixed size like this.

File: .gitignore
Comment (johnyenter-briars) [reply .gitignore:219]: :+1:

**Comment by johnyenter-briars** [comment]
Ok cool! I'm glad you think this is something worth pursuing. 

I will move forward on making it a setting option. (Probably a checkbox? If it's a boolean setting? Either permanent icon size or not)

To clarify, it looks like we could take this one of two ways. Which option set do you feel I should target?

#### Option Set 1
- The user could chose to have the album pixmap size static (64 pix) or dynamic (dynamic in the same way the artist cover is)

or

#### Option Set 2
- The user could chose to have the album pixmap size static (64 pix) or "large" (maybe something on the order of 300)  but still static. 

I will also start work on updating `/view/context/content.cpp` to support either a scalable album size or the default size.

Also full disclosure: c++ is not my native language, so I'm relying on what I remember from university to get me through :)

File: src/mainwindow.cpp
Comment (johnyenter-briars) [reply src/mainwindow.cpp:423]: Yea, I did this because I added an overload method `setAlbumJpg` farther down the call tree that takes in a string and converts it to a pix map. But I think with your suggestions this change not be needed, if `Context::Content::Content` can have like a `scaleSize` method to update it's album art's size.

Comment (johnyenter-briars) [reply src/mainwindow.cpp:423]: Yea, I did this because I added a method setAlbumJpg farther down the call tree, and so I added a bunch of overloaded methods on the way up. Agreed, that's probably not the best way to do it - lots of code duplication for no good reason.

But I think with your suggestions this change will not be needed, if Context::Content::Content can have like a scaleSize method to update it's album art's size.

I'm hoping to refactor in such a way so that way I can revert this change.

**Comment by kraxarn** [comment]
I think option 1 is the better one, just makes more sense to me.
The problem I guess with `Context::Content` is that it makes use of an horizontal layout, instead of a vertical one that would be needed here. Because they will look, and behave, quite differently, maybe making an entirely new widget for it could work? `Context::NowPlaying` could still be useful for example.

Maybe an "expand album cover" option? Settings can be added to this: 
https://github.com/kraxarn/spotify-qt/blob/b4dbf9b1bf6481db5782a9a528d7caf045f5a22b/lib/include/lib/settings/qt.hpp#L12
And a checkbox can be added to:
https://github.com/kraxarn/spotify-qt/blob/b4dbf9b1bf6481db5782a9a528d7caf045f5a22b/src/settingspage/interface.cpp#L14

**Comment by johnyenter-briars** [comment]
> I think option 1 is the better one, just makes more sense to me. The problem I guess with `Context::Content` is that it makes use of an horizontal layout, instead of a vertical one that would be needed here. Because they will look, and behave, quite differently, maybe making an entirely new widget for it could work? `Context::NowPlaying` could still be useful for example.
> 
> Maybe an "expand album cover" option? Settings can be added to this:
> 
> https://github.com/kraxarn/spotify-qt/blob/b4dbf9b1bf6481db5782a9a528d7caf045f5a22b/lib/include/lib/settings/qt.hpp#L12
> 
> And a checkbox can be added to:
> 
> https://github.com/kraxarn/spotify-qt/blob/b4dbf9b1bf6481db5782a9a528d7caf045f5a22b/src/settingspage/interface.cpp#L14

Ok! I'll start working on these features and report back once I have something relevant to show. : )

**Comment by kraxarn** [comment]
I have now added support for multiple album covers in d3f7b261bf154a17b09759d275b8647d680ee0a2, but most importantly, there's now `lib::spt::track::image_small()` and `lib::spt::track::image_large()` for getting the small and large album covers respectively.

**Comment by johnyenter-briars** [comment]
Hey @kraxarn ! 

I'm back again. Would you mind taking a look at what I've got so far?

So the basic idea is that I just added another class similar to `Content` but can horizontally scale. I then updated a bunch of the other methods which interact with that QWidget to account for the updated logic.

I attached a video of the new QWidget in action (sorry if it's a little cut off):
https://user-images.githubusercontent.com/42982326/152408062-cd077fcc-e050-44ff-a153-25fc41d99b07.mp4

I think this implementation isn't quite right though, as I'm getting occasional segfaults when trying to switch between the two options in the settings.

And this is where my knowledge of C++ is failing me. I thought the segfaults had to do with the the fact that I was freeing objects that still had dependent objects on the heap. So I tried to add destructors to each dependent object to each free their own data. But I'm still getting segfaults. I will mark in the file exactly where each segfault is.

The other approach I tried was to build an abstract implementation of "content", `AbstractContent` which would inherit from QWidget, and then both `Content` and `HorizContent` would inherit from `AbstractContent`. And that way the `View` wouldn't have to manage two instances of "content", it would just have the one abstract implementation.  

Unfortunately I couldn't get this to work either - the code would compile, but then the linker would yell at me something about a vtable. 

Do you have any thoughts on this approach? Is there another way I should attack this? Any advice would be appreciated.

File: src/view/context/view.cpp
Comment (johnyenter-briars) [inline src/view/context/view.cpp:36]: Will occasionally seg fault here.

File: src/view/context/view.cpp
Comment (johnyenter-briars) [inline src/view/context/view.cpp:44]: Will occasionally seg fault here.

File: src/view/context/view.cpp
Comment (johnyenter-briars) [inline src/view/context/view.cpp:48]: Will occasionally seg fault here.

File: src/view/context/view.cpp
Comment (johnyenter-briars) [inline src/view/context/view.cpp:38]: Like I feel like this is not the optimal way to perform object updates. It just feels wrong - but IDK how else to do something like this in C++ :pleading_face:

**Comment by johnyenter-briars** [comment]
Ok - now I think the segfauls are gone. Needed to use `delete` instead of `free` : )

File: src/view/context/view.cpp
Comment (kraxarn) [reply src/view/context/view.cpp]: `free`, and `malloc` to allocate memory, is the "C way" of doing memory management, while C++ prefers using `delete` and `new` respectively. However, in Qt, you very rarely need to do manual memory management. If you manually need to delete something, like when switching out a widget or similar, you can `QObject::deleteLater` instead.

**Comment by johnyenter-briars** [comment]
Hey there @kraxarn !

Would you mind taking a look at the code when you get a chance please? 

I updated the code to use `QObject::deleteLater` instead of a destructor. 

The functionality is still the exact same as demoed [here](https://user-images.githubusercontent.com/42982326/152408062-cd077fcc-e050-44ff-a153-25fc41d99b07.mp4)

File: src/view/context/albumcover.cpp
Comment (kraxarn) [inline src/view/context/albumcover.cpp:2]: Avoid "using namespace", but this one doesn't seem to be used anyway

File: src/view/context/view.cpp
Comment (kraxarn) [inline src/view/context/view.cpp:35]: You don't have to call `QObject::deleteLater()` from within the instance itself. It's like you're overriding the method, but just call the base method anyway, not really doing anything. `content->deleteLater()` does the same thing.

File: src/view/context/view.cpp
Comment (kraxarn) [inline src/view/context/view.cpp:88]: This is fine, but as we're in Qt, we can also use `qFatal()`, which more or less does the same thing, just the "Qt-way".

File: src/view/context/view.cpp
Comment (kraxarn) [inline src/view/context/view.cpp:93]: Having a common interface for both of these would make this a lot prettier, but it's not a requirement. Did you try using pure virtual methods?

File: src/view/context/horizcontent.hpp
Comment (kraxarn) [inline src/view/context/horizcontent.hpp:42]: Unused, remove :p

File: src/view/context/nowplaying.cpp
Comment (kraxarn) [inline src/view/context/nowplaying.cpp:64]: All of these have a parent and don't really need to be manually freed.

**Review by kraxarn — CHANGES_REQUESTED** [review]
Just a few more changes before I'm ready to merge :relaxed:

File: src/view/context/view.cpp
Comment (johnyenter-briars) [reply src/view/context/view.cpp:35]: Oh DUH! Yea that's a stoopid miss on my end :)

File: src/view/context/view.cpp
Comment (johnyenter-briars) [reply src/view/context/view.cpp:93]: Yea - I tried to do this using [this stackoverflow post](https://stackoverflow.com/questions/9756893/how-to-implement-interfaces-in-c) but got a ton of obtuse linker errors. The code would compile - but then the linker would complain. Something about "vtable not found".

I'll give it another shot!

Comment (kraxarn) [reply src/view/context/view.cpp:93]: Don't know how helpful it is, but there's a similar thing with paths, where I have a generic interface for the methods:
https://github.com/kraxarn/spotify-qt/blob/894e220356183344043b74d43ab696a04e21ed25/lib/include/lib/paths/paths.hpp#L13
And then an implementation for Qt:
https://github.com/kraxarn/spotify-qt/blob/894e220356183344043b74d43ab696a04e21ed25/src/lib/qtpaths.hpp#L8

**Comment by johnyenter-briars** [comment]
Hey @kraxarn ! Me again! Sorry it took me so long to get this done. Man C++ is hard : ) ! 

Would you mind taking a look at where we're at right now when you get the chance please?

**Review by kraxarn — CHANGES_REQUESTED** [review]
C++ can be difficult if you aren't used to it :p
Almost ready to merge now :smile:

File: src/view/context/abstractcontent.hpp
Comment (johnyenter-briars) [reply src/view/context/abstractcontent.hpp:32]: cause also, i think `setCurrentlyPlaying` and `onSongMenu` can also be moved to abstractcontent.cpp

Comment (kraxarn) [reply src/view/context/abstractcontent.hpp]: Sorry for the slow reply.

I think trying to share as much as possible between them makes more sense, and having a constructor that just passes most values on to the base constructor isn't a bad idea at all. `resizeEvent` makes sense that they differ, but isn't `setAlbum` the same between them? It just sets a pixmap and tooltip, which should be the same in both.

Comment (johnyenter-briars) [reply src/view/context/abstractcontent.hpp]: > Sorry for the slow reply sweat_smile
> 
> I think trying to share as much as possible between them makes more sense, and having a constructor that just passes most values on to the base constructor isn't a bad idea at all. `resizeEvent` makes sense that they differ, but isn't `setAlbum` the same between them? It just sets a pixmap and tooltip, which should be the same in both.

Ok cool - I kinda figured that we might go with the "have as much shared as possible", so I have a extra branch on my local machine that has everything moved over. I will merge it into `album_size` once you say yay or nay on `setAlbum` : )

Regarding `setAlbum`, you're right - they're nearly identical, the only difference is that the one in just `Content` places the mask.

content.cpp
```c++
void Context::Content::setAlbum(const lib::spt::entity &albumEntity, const QPixmap &albumImage)
{
	if (album != nullptr)
	{
		album->setPixmap(Image::mask(albumImage));
		album->setToolTip(QString::fromStdString(albumEntity.name));
	}
}
```

horizcontent.cpp
```c++
void Context::HorizContent::setAlbum(const lib::spt::entity &albumEntity, const QPixmap &albumImage)
{
	if (album != nullptr)
	{
		album->setPixmap(albumImage);
		album->setToolTip(QString::fromStdString(albumEntity.name));
	}
}
```
So I wasn't quite sure what to do with this one. I could add a parameter flag thats just like `addImageMask` or something like that. But I wanted to know if you had any better ideas.

Comment (kraxarn) [reply src/view/context/abstractcontent.hpp]: Any reason to not use `Image::mask` in both cases? I plan on adding a setting to use the mask or not, but in the meantime, I think just using it for both makes the most sense.

Comment (johnyenter-briars) [reply src/view/context/abstractcontent.hpp]: Ok - adding the mask in will make album corners be cut off - even while expandable. 

image

**Comment by johnyenter-briars** [comment]
@kraxarn - would you mind taking a look when you get a chance please? Like we discussed in the above thread - this commit should cover those methods being moved into `AbstractContent`

File: src/view/context/abstractcontent.hpp
Comment (kraxarn) [reply src/view/context/abstractcontent.hpp]: It's kind of the point, but again, adding an option to disable it doesn't hurt. Guess it's much more noticeable with a larger image.

**Comment by kraxarn** [comment]
It's only a thing I've been thinking about doing, as well as possibly adding some other shapes to choose from, but haven't implemented anything for it yet. If you want to take a look at it, you are very welcome to :relaxed:

Either way, I'm ready to merge this when you are! Thanks a lot for your help and patience :smile:

**Review by kraxarn — APPROVED** [review]

**Comment by johnyenter-briars** [comment]
> It's only a thing I've been thinking about doing, as well as possibly adding some other shapes to choose from, but haven't implemented anything for it yet. If you want to take a look at it, you are very welcome to relaxed
> 
> Either way, I'm ready to merge this when you are! Thanks a lot for your help and patience smile

No problem! Thank you for bearing with my c++ "noobness" -  I hope people find the feature useful! Marked as "ready for review" :smile:

**Pull request merged**
